[fix] #2227: Implement Register and Unregister for Asset

QuentinI opened the pull request:
<!-- You will not see HTML commented line in Pull Request body -->
<!-- Optional sections may be omitted. Just remove them or write None -->

<!-- ### Requirements -->
<!-- * Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion. -->
<!-- * All new code must have code coverage above 70% (https://docs.codecov.io/docs/about-code-coverage). -->
<!-- * CircleCI builds must be passed. -->
<!-- * Critical and blocker issues reported by Sorabot must be fixed. -->
<!-- * Branch must be rebased onto base branch (https://soramitsu.atlassian.net/wiki/spaces/IS/pages/11173889/Rebase+and+merge+guide). -->


### Description of the Change

- Implemented Register and Unregister instructions for Asset
- Changed minting implementation to require registering the asset first
- Adjusted public validators accordingly

<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. -->
<!-- Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->

### Issue

Resolves #2227 

<!-- If it is not a GitHub issue but a JIRA issue, just put the link here -->

### Benefits

Consistency, no more "unsupported" Register/Unregister instructions

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks

Register asset and Mint should now be considered together when implementing permission systems.

<!-- What are the possible side-effects or negative impacts of the code change? -->
<!-- If no drawbacks, explicitly mention this (write None) -->


<!--
NOTE: User may want skip pull request and push workflows with [skip ci]
https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/
Phrases: [skip ci], [ci skip], [no ci], [skip actions], or [actions skip]
-->

mversic commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -54,6 +54,59 @@ pub mod isi {
     }
 
+    #[allow(clippy::panic_in_result_fn)]
+    impl<W: WorldTrait> Execute<W> for Register<Asset> {
```
also, please compare with [this](https://github.com/hyperledger/iroha/pull/2330)

QuentinI commented on cli/src/torii/tests.rs:
```diff
@@ -278,8 +278,12 @@ fn register_asset_definition(name: &str) -> Instruction {
+fn register_asset(quantity: u32, asset: &str, account: &str) -> Instruction {
+    RegisterBox::new(Asset::new(
+        asset_id(asset, account),
+        AssetValue::Quantity(quantity),
```
Oh, thanks. I didn't know it was possible, I'll fix it everywhere.

mversic commented on configs/peer/genesis.json:
```diff
@@ -62,6 +62,35 @@
+          }
+        },
         {
           "Mint": {
```
Ok, this was my dilema as well. I don't see a downside

We could still remove the mint instruction from genesis, though

mversic replied:
The problem is it's implicit. It might be confusing to the user

QuentinI replied:
Sure, I'll create one.
I think I'll leave it as Mint + Register for now and will then change it depending on what we decide.

QuentinI replied:
Ah, I see it's already there.

outoftardis commented on permissions_validators/src/public_blockchain/burn.rs:
```diff
@@ -5,7 +5,7 @@ use iroha_data_model::asset::DefinitionId;
 
 declare_token!(
-    /// Can burn asset with the corresponding asset definition.
+    /// Can burn and unregister asset with the corresponding asset definition.
```
```suggestion
    /// Can burn and unregister assets with the corresponding asset definition.
```

outoftardis commented on permissions_validators/src/public_blockchain/mint.rs:
```diff
@@ -4,7 +4,7 @@ use iroha_data_model::asset::DefinitionId;
 
 declare_token!(
-    /// Can mint asset with the corresponding asset definition.
+    /// Can register and mint asset with the corresponding asset definition.
```
```suggestion
    /// Can register and mint assets with the corresponding asset definition.
```

outoftardis commented on core/src/wsv.rs:
```diff
@@ -296,6 +296,7 @@ impl<W: WorldTrait> WorldStateView<W> {
     /// # Errors
     /// - No such [`Asset`]
     /// - No [`Account`] to which the [`Asset`] is associated.
+    /// - No [`Domain`] to which the [`Account`] is associated.
```
```suggestion
    /// - No [`Account`] with which the [`Asset`] is associated.
    /// - No [`Domain`] with which the [`Account`] is associated.
```

outoftardis commented on permissions_validators/src/public_blockchain/burn.rs:
```diff
@@ -70,22 +95,33 @@ impl<W: WorldTrait> HasToken<W> for GrantedByAssetCreator {
+                {
+                    Ok(CanBurnAssetWithDefinition::new(asset_id.definition_id).into())
+                } else {
+                    Err("Instruction is not unregister asset".to_owned())
```
here and in other similar error messages in this PR: I'd suggest saying what was expected, e.g. something like `Expected the unregister instruction` or `Expected the unregister asset instruction`

appetrosyan reviewed (commented):
This change is important for SDK developers. Need to update with #api_changes.

QuentinI commented on tools/kagami/src/main.rs:
```diff
@@ -186,6 +186,14 @@ mod genesis {
             .with_asset("rose".parse()?, AssetValueType::Quantity)
             .finish_domain()
             .build();
+        let register = RegisterBox::new(iroha_core::tx::Asset::new(
```
Will do

Arjentix commented on client/tests/integration/asset.rs:
```diff
@@ -9,6 +9,77 @@ use test_network::*;
+        })
+    })?;
+
+    let hash = test_client.submit(register_asset)?;
+
+    // But registering an asset to account already having one should fail
+    test_client.poll_request(client::transaction::by_hash(*hash), |tx| {
+        matches!(tx, TransactionValue::RejectedTransaction(_))
+    })?;
```
```suggestion
    // But registering an asset to account already having one should fail
    assert!(test_client.submit_blocking(register_asset).is_err());
```

Arjentix replied:
Should work the same but looks easier

Arjentix commented on core/src/smartcontracts/isi/account.rs:
```diff
@@ -15,6 +15,59 @@ use crate::{ValidQuery, WorldStateView};
+            let asset_id = self.object.id().clone();
+
+            match wsv.asset(&asset_id) {
+                Err(e) if matches!(e, FindError::Asset(_)) => {
```
```suggestion
                Err(FindError::Asset(_)) => {
```

QuentinI replied:
🤦 
Right, thanks

Arjentix commented on core/src/smartcontracts/isi/account.rs:
```diff
@@ -15,6 +15,59 @@ use crate::{ValidQuery, WorldStateView};
+                Err(e) if matches!(e, FindError::Asset(_)) => {
+                    let res = wsv.modify_account(&asset_id.account_id, |account| {
+                        // We are in an arm where asset doesn't exist
+                        assert!(account.add_asset(self.object).is_none());
```
What's the difference with `unwrap()`? Style? 🤔

QuentinI replied:
It seems to me that it's more appropriate semantically, since it emphasizes that this should be unreachable state, not just unrecoverable. `unreachable!` would be even better semantically, but it would look awkward.

Arjentix commented on core/src/smartcontracts/isi/account.rs:
```diff
@@ -15,6 +15,59 @@ use crate::{ValidQuery, WorldStateView};
+            let asset_id = self.object_id;
+            let account_id = asset_id.account_id.clone();
+
+            wsv.modify_account(&account_id, |acct| {
```
I'd like more clear name than more cryptographic one

Arjentix commented on core/src/smartcontracts/isi/mod.rs:
```diff
@@ -263,11 +263,11 @@ impl<W: WorldTrait> Execute<W> for RegisterBox {
                 Register::<Trigger<FilterBox>>::new(*trigger).execute(authority, wsv)
             }
             RegistrableBox::Role(role) => Register::<Role>::new(*role).execute(authority, wsv),
-            _ => Err(Error::Unsupported(InstructionType::Register)),
```
That's cool. Now we will have a compile-time error if forget to add branch for something new

appetrosyan replied:
Technically, most of our wildcards could be removed and our code would become more reliable.

QuentinI commented on client/tests/integration/unstable_network.rs:
```diff
@@ -46,7 +46,7 @@ fn unstable_network_7_peers_1_fault() {
 
 #[test]
-#[ignore = "This test does not guarantee to have positive outcome given a fixed time."]
+// #[ignore = "This test does not guarantee to have positive outcome given a fixed time."]
```
Oh, it's an accident

Arjentix commented:
I'll approve the when all tests will pass

QuentinI commented:
@Arjentix I don't think it's ready yet. Apart from there not being any consensus on #2339 , this PR as it is now introduces a loophole to mint non-mintable assets twice.

Arjentix commented:
Oh, yes, haven't think about it...

appetrosyan commented:
Overall, I think we need to create an integration test for the double mint loophole. To me it seems that this could be fixed by pattern-matching on the initial amount, and registering the asset as `Mintable::Not` right after its initial registration. 

So if you want to defer the minting, you can still do it, but if all you want is to create a non-mintable asset with an initial quantity, you can do that too.

outoftardis commented on core/src/wsv.rs:
```diff
@@ -295,7 +295,8 @@ impl<W: WorldTrait> WorldStateView<W> {
     /// # Errors
     /// - No such [`Asset`]
-    /// - No [`Account`] to which the [`Asset`] is associated.
+    /// - No [`Account`] with which the [`Asset`] is associated.
+    /// - No [`Domain`] with which the [`Account`] is associated.
```
```suggestion
    /// - The [`Asset`] is not associated with any [`Account`].
    /// - The [`Account`] is not associated with any [`Domain`].
```

QuentinI replied:
It's not that it's not associated with any account/domain, it's that specific account/domain it's supposed to be associated with doesn't exist.

Would "Assotiated X doesn't exist" be better?

appetrosyan replied:
```suggestion
    /// - The [`Account`] with which the [`Asset`] is associated doesn't exist. 
    /// - The [`Domain`] with which the [`Account`] is associated doesn't exist.
```

outoftardis commented on core/src/genesis.rs:
```diff
@@ -476,6 +502,28 @@ impl RawGenesisDomainBuilder {
+        }
+    }
+
+    /// Add [`Asset`] to current account initialized with the given value
```
```suggestion
    /// Add [`Asset`] to current account and initialize it with the given value
```

outoftardis commented on core/src/genesis.rs:
```diff
@@ -476,6 +502,28 @@ impl RawGenesisDomainBuilder {
 }
 
+impl RawGenesisAccountBuilder {
+    /// Finish this account and return to
```
```suggestion
    /// Finish building this account and return to
```

outoftardis commented on core/src/genesis.rs:
```diff
@@ -443,25 +450,44 @@ impl RawGenesisDomainBuilder {
     }
 
-    /// Add an account to this domain
+    /// Add an account to this domain and return account builder
+    /// which can be used to register assets to the account.
```
```suggestion
    /// Add an account to this domain and return an account builder,
    /// which can be used to register assets to this account.
```

QuentinI commented:
Based on #2339 I removed requirement to register asset before minting and removed `Mintable::Once`.

I experimented with default values in asset definitions, but it ultimately resulted in convoluted API when you take NFTs into account, so I scrapped that. I still think asset registration semantics as implemented here is not ideal, so I would maybe like to try and revisit that after release.

QuentinI commented:
> Overall, I think we need to create an integration test for the double mint loophole. To me it seems that this could be fixed by pattern-matching on the initial amount, and registering the asset as `Mintable::Not` right after its initial registration.
> 
> So if you want to defer the minting, you can still do it, but if all you want is to create a non-mintable asset with an initial quantity, you can do that too.

@appetrosyan sorry, I missed your comment.
I think disallowing minting altogether and leaving registration as the only way to add non-mintable assets to accounts results in more straightforward API. Is there a particular reason you would want to support minting in this case?

appetrosyan commented on client/tests/integration/non_mintable.rs:
```diff
@@ -8,30 +8,27 @@ use iroha_data_model::{metadata::UnlimitedMetadata, prelude::*};
     let asset_definition_id = AssetDefinitionId::from_str("xor#wonderland").expect("Valid");
     let create_asset =
-        RegisterBox::new(AssetDefinition::quantity(asset_definition_id.clone()).mintable_once());
+        RegisterBox::new(AssetDefinition::quantity(asset_definition_id.clone()).not_mintable());
```
I don't see how `mintable_once` is problematic, in this context: these assets can really be minted once.

QuentinI replied:
After this PR they become actually non-mintable, only registrable. See my other [comment](https://github.com/hyperledger/iroha/pull/2333#discussion_r900331226)

appetrosyan commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -16,6 +16,53 @@ use super::prelude::*;
+        }
+    }
+
+    impl Execute for Unregister<Asset> {
```
This is an important point for SDK developers.

appetrosyan commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -318,15 +365,6 @@ pub mod isi {
         match definition.mintable() {
```
At this point, we could revert back to a boolean. But I would recommend keeping the `Mintable::Once` variant. We want to offer people the choice to register an asset definition but to defer the minting.

QuentinI replied:
You can still register _asset definition_ without minting.
After this PR `Mintable::Not` behaves essentially the same as `Mintable::Once` currently, since you can register _asset to an account_ with whatever value, that's why I removed it.

> At this point, we could revert back to a boolean

Didn't do it based on [this](https://github.com/hyperledger/iroha/blob/916a2cc584ba110775eafeae69e9658157f704c9/data_model/src/asset.rs#L191)

pesterev commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -16,6 +16,55 @@ use super::prelude::*;
+                Err(e) if matches!(e, FindError::Asset(_)) => {
+                    assert_can_register(&asset_id.definition_id, wsv, self.object.value())?;
+
+                    if wsv
+                        .asset_or_insert(asset_id, self.object.value().clone())
+                        .is_err()
+                    {
+                        unreachable!("Account is checked to exist")
+                    }
```
```suggestion
                    wsv.asset_or_insert(asset_id, self.object.value().clone()).expect("Account exists");
```
I think it should be replaced with `expect()` so that the panic matches the semantics and explains what is expected here.

appetrosyan commented:
> I think disallowing minting altogether and leaving registration as the only way to add non-mintable assets to accounts results in more straightforward API. Is there a particular reason you would want to support minting in this case?

The APIs are comparably simple, but in different ways. 

If you retained the `Mintable::Once` variant, you can create an asset and can defer the amount that you wanted to mint. Why is this good? Consider that we made `Mint` variadic: it accepts multiple `(AccountId, Asset, Quantity)` tuples, and so if you want to create a non-mintable asset that is split evenly between e.g. shareholders of a company, you can do that in two instructions. If we don't allow it, we increase the load on the network, because to achieve the same result the person doing it has to split the single `Mint` ISI into a `Mint` and several transfers. 

Moreover, we will need to synchronise the behaviour of `Mint` with the implementation in `Execute for Register<AssetDefinition>`, so nominally you have fewer instructions but each of them is more complex. 

Both approaches are valid: in the CPU world what you propose is called CISC, while my stance is closer to RISC with SIMD. I personally would prefer that we kept the API changes in this PR to a minimum, but will approve given a good justification.

QuentinI commented:
Right, my stance on this was the way it was mostly because I confused AssetDefinition and Asset in regards to Mintable.
I reverted to having Mintable::Once and implemented your proposed solution for double-mint loophole:

> So  if Register has value 0, creates a Mintable::Once asset, while a register with any other non-zero value necessarily a Mintable::Not variant

I have one question in regards to that - in this implementation one can register Mintable::Once asset to any number of accounts as long as Register has value 0. Are we okay with that?

mversic commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -16,6 +16,50 @@ use super::prelude::*;
     use super::*;
 
+    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    impl Execute for Register<Asset> {
```
`Register<Asset>` and `Unregister<Asset>` should be placed in `smartcontracts/account.rs` since that seems to be the convention

QuentinI replied:
Right, I guess I messed up while rebasing and it ended back there.

mversic commented on core/src/smartcontracts/isi/asset.rs:
```diff
@@ -319,12 +376,26 @@ pub mod isi {
             Mintable::Infinitely => Ok(()),
             Mintable::Not => Err(Error::Mintability(MintabilityError::MintUnmintable)),
             Mintable::Once => {
```
if we'll allow for initial value to be set in the asset registration request (as far as I see this is the case) then I don't see a need for `Mintable` to be enum, rather it can be a `bool` since there is no need for `Mintable::Once` any longer

SamHSmith reviewed (approved):
LGTM. But you should probably address marin's comment.
